### PR TITLE
[WIP] Wrap HashBucket functionality into Categorify op

### PR DESCRIPTION
**Pull request by rnyak**
This PR is created 
- to address FEA https://github.com/NVIDIA/NVTabular/issues/78 : currently it can perform hashing when calling categorify op providing freq_threshold = 0 and num_buckets as kwargs.  The hashing case where freq_threshold >1 is not applied yet.

### CODE REVIEW

**Comment by rnyak** [comment]
> These changes look good @rnyak - Thanks!
> 
> I think the only thing missing is a simple test to confirm that the encoding works for validation data with new categories (the motivation for this PR). It would also be best to test the `search_sorted=True` case now that it is not the default (otherwise we are losing test coverage for that algorithm).

@rjzamora do you mean adding a unit test?  if yes, sure I will add that. Also,  since I am wrapping hash_bucket functionality into categorify the code will change again.

**Comment by rjzamora** [comment]
> @rjzamora do you mean adding a unit test? if yes, sure I will add that. Also, since I am wrapping hash_bucket functionality into categorify the code will change again.

Yep, exactly.  With every PR we want to make sure that (1) the unit tests are covering new functionality, and (2) the unit tests are still covering the functionality from before the PR.

File: nvtabular/ops/categorify.py [hunk 1]
```diff
@@ -740,7 +747,7 @@ def _encode(name, storage_name, path, gdf, cat_cache, na_sentinel=-1, freq_thres
         value.index.name = "labels"
         value.reset_index(drop=False, inplace=True)
```
Comment (rjzamora) [inline nvtabular/ops/categorify.py:749]: May need to raise an error if `freq_threshold>0` **and**  `search_sorted` is `True`.  However, it is probably best to catch this problem quickly in the `Categorify` initializer.

File: nvtabular/ops/categorify.py [hunk 2]
```diff
@@ -753,8 +773,8 @@ def _encode(name, storage_name, path, gdf, cat_cache, na_sentinel=-1, freq_thres
         ).sort_values("order")["labels"]
         labels.fillna(na_sentinel, inplace=True)
         labels = labels.values
-    else:
-        # Use `searchsorted` if we are using a "full" encoding
+    if search_sorted is True:
```
Comment (benfred) [inline nvtabular/ops/categorify.py:776]:
```suggestion
    if search_sorted:
```

File: nvtabular/ops/categorify.py [hunk 3]
```diff
@@ -740,7 +760,7 @@ def _encode(name, storage_name, path, gdf, cat_cache, na_sentinel=-1, freq_thres
         value.index.name = "labels"
         value.reset_index(drop=False, inplace=True)
 
-    if freq_threshold > 0:
+    if search_sorted is False or freq_threshold > 0:
```
Comment (benfred) [inline nvtabular/ops/categorify.py:763]:
```suggestion
    if not search_sorted or freq_threshold > 0:
```

File: nvtabular/ops/categorify.py [hunk 4]
```diff
@@ -764,7 +784,8 @@ def _encode(name, storage_name, path, gdf, cat_cache, na_sentinel=-1, freq_thres
                 gdf[selection_l], side="left", na_position="first"
             )
         labels[labels >= len(value[selection_r])] = na_sentinel
-
+    if search_sorted is True and freq_threshold > 0:
```
Comment (benfred) [inline nvtabular/ops/categorify.py:787]:
```suggestion
    if search_sorted and freq_threshold > 0:
```

File: nvtabular/ops/categorify.py [hunk 5]
```diff
@@ -764,7 +784,8 @@ def _encode(name, storage_name, path, gdf, cat_cache, na_sentinel=-1, freq_thres
...
+        raise ValueError("cannot use search_sorted=True with freq_threshold > 0.")
```
Comment (benfred) [inline nvtabular/ops/categorify.py:788]: We probably want this check in the constructor - I think at this stage we will have already calculated the categories and applied them to at least one partition before failing,

File: nvtabular/ops/categorify.py [hunk 6]
```diff
@@ -277,7 +285,17 @@ def apply_op(
                 new_gdf[name] = new_gdf[new_col]
                 new_gdf.drop(columns=[new_col], inplace=True)
 
+        if self.num_buckets is not None and self.freq_threshold == 0:
+            hash_bucket_op = HashBucket(self.num_buckets)
+            new_gdf = hash_bucket_op.apply_op(new_gdf, columns_ctx, "categorical")
+
+        if self.num_buckets == 0:
+            raise ValueError(
+                "For hashing num_buckets should be an int > 1, otherwise set it to None."
```
Comment (benfred) [inline nvtabular/ops/categorify.py:294]: We should move this check to the consturctor

File: nvtabular/ops/categorify.py [hunk 7]
```diff
@@ -277,7 +285,17 @@ def apply_op(
                 new_gdf[name] = new_gdf[new_col]
                 new_gdf.drop(columns=[new_col], inplace=True)
 
+        if self.num_buckets is not None and self.freq_threshold == 0:
+            hash_bucket_op = HashBucket(self.num_buckets)
+            new_gdf = hash_bucket_op.apply_op(new_gdf, columns_ctx, "categorical")
```
Comment (benfred) [inline nvtabular/ops/categorify.py:290]: Whats the use case for doing the full categorical encoding and then hashing the output?  (like  do we need to incur the expense of doign the categorical encoding first if we are just going to hash everything after?)

File: nvtabular/ops/categorify.py (hunk 5 above)
Comment (rnyak) [reply nvtabular/ops/categorify.py:788]: @benfred thanks, yes Rick also mentioned that, so I added this condition into the constructor. will push the changes.

File: nvtabular/ops/categorify.py (hunk 7 above)
Comment (rnyak) [reply nvtabular/ops/categorify.py:290]: @benfred good point. you are right. I am working on that part. The example for that would be the case where `freq_threshold >1 + num_buckets >1`.

**Comment by rnyak** [comment]
@benfred closing this will create a new one.

**Pull request closed without merging**